ovs-dpdk.md

```diff
@@ -0,0 +1,180 @@
+
+
+## Configure OVS and DPDK on the Host
+
+### Install dpdk packages:
+
+```bash
+# yum install dpdk dpdk-tools driverctl
+
+```
+ 
+### Bind NICs to DPDK:
+
+In order to run a DPDK application, a suitable uio module should be loaded into kernel memory.
+Available modules are: uio_pci_generic, igb_uio and vfio-pci.
+In many cases, the standard uio_pci_generic module included in the Linux kernel can provide the uio capability.
+For some devices which lack support for legacy interrupts, igb_uio should be used, but it’s an out-of-tree kernel module.
+VFIO is preferred to uio drivers in latest DPDK versions in platforms that support VFIO.
+
+All ports that you would like to run DPDK application should be bound to one of the modules before OVS is started. DPDK will ignore all devices which are not bound to one of the drivers above.
+ 
+Load vfio-pci
+
+```
+# modprobe vfio_pci
+```
+Find the appriopriate network devices you wish to bind:
+```
+# lspci | grep Ether
+```
+Review the status of current port bindings:
+```
+# driverctl -v list-devices
+```
+Bind the device to DPDK:
+```
+# driverctl set-override <PCI_ID> vfio-pci
+```
+
+### Hugepages Support
+Hugepages support is required for increasing performance, they allow the Linux kernel to manage larger pages, in addition to the standard 4K pages.
+Modern cpus use page-based mechanism to translate virtual address to physical address. The mapping is stored in page tables and can grow significantly large and to make the translation faster we use a transparant cache called TLB (Translation Lookaside Buffer). TLB miss will be handled by the hardware, which will increase the translation time significantly. TLB misses can be reduced by increasing the page size. Therefore, 1GiB pages are preferable.
+
+Number of pages can vary according to usage requirements, but is limited to memory size.
+Hugepages must be marked as shared in order to enable host and guest communication, as packets are transmitted and received on buffers allocated on hugepages memory. 
+
+Please note that 1Gib hugepages feature is not supported on every CPU 
+
+Review custom properties
+```
+# engine-config -g UserDefinedVMProperties
+```
+Add engine custom properties for hugepages support (followed by engine restart)
+```
+# engine-config -s "UserDefinedVMProperties=hugepages=^.*$;hugepages_shared=(true|false)"
+```
+ 
+
+### Enable DPDK in OVS:
+
+ovs-vscwhitchd should get a few parameters in order to enable DPDK functionality.
+
+* Set ```dpdk-init``` so OVS would initialize and support DPDK ports:
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-init=true
+```
+ 
+
+* ```dpdk-socket-mem```: a comma-separated list of amount of memory to allocate from each NUMA node. It is important to allocate hugepage memory to all NUMA nodes that have DPDK interfaces associated with them, otherwise, the NUMA node cannot be used and errors will occur. 
```

Comment:
Suggest:
": A comma-separated list of the amount of memory.."
"It is important to allocate huge page memory.."